File: apps/api/src/modules/trails/domain/trail.ts
import type { RouteBounds, RouteGeometry } from '../../routing/domain/route.js';

/**
 * Trail domain types (§6).
 *
 * A Trail is **not** a Route (constitution §Trail). A Route is what the road network
 * says about getting from A to B — geometry, distance, duration, computed on demand
 * and owned by nobody. A Trail is what a person decided: which places, in which order,
 * deliberately saved. Routing answers *how*; a Trail records *what was chosen*.
 *
 * Nothing here is social. No visibility, no slug, no counts: publication is a later
 * PR with rules of its own, and a field reserved for it now is an invitation to fill
 * it with something that has none (§72).
 */

export const TrailStatus = {
  /** Being composed. The only state a Trail is created in. */
  DRAFT: 'DRAFT',
  /** The user said they were done. Not published — nobody else can see it (§45). */
  FINALIZED: 'FINALIZED',
  /** Finished with, kept but not listed by default. */
  ARCHIVED: 'ARCHIVED',
} as const;

export type TrailStatus = (typeof TrailStatus)[keyof typeof TrailStatus];

/**
 * Where a stop came from (§11).
 *
 * Provenance of the composition, never a quality signal and never an input to a
 * ranking. `AI` is absent because nothing can produce it yet.
 */
export const TrailStopSource = {
  DISCOVERY: 'DISCOVERY',
  SEARCH: 'SEARCH',
  MANUAL: 'MANUAL',
} as const;

export type TrailStopSource = (typeof TrailStopSource)[keyof typeof TrailStopSource];

export const TRAIL_STOP_SOURCES: readonly TrailStopSource[] = Object.values(TrailStopSource);

/**
 * One end of a Trail, snapshotted onto it (§9).
 *
 * A Trail's endpoints need not be Places — "from where I am now" is a perfectly good
 * origin — so the coordinate is the truth and `placeId` is recorded only when one was
 * chosen. `label` preserves what the user was looking at when they chose it.
 */
export interface TrailEndpoint {
  readonly latitude: number;
  readonly longitude: number;
  readonly placeId: string | null;
  readonly label: string | null;
}

/**
 * A stop, which is always a resolved Place (constitution §Domain, §10).
 *
 * Never a loose coordinate and never free text: an unresolved stop cannot be shown,
 * searched, deduplicated or reused across Trails, and admitting one would make every
 * consumer handle a case that should not exist.
 */
export interface TrailStop {
  readonly id: string;
  readonly placeId: string;
  /** 1-based and contiguous. Never derived from insertion time (§12). */
  readonly position: number;
  readonly source: TrailStopSource;
  /** Denormalised for display only; the Place remains the source of truth (§52). */
  readonly placeName: string;
  readonly placeCategoryId: string;
  readonly placeLatitude: number;
  readonly placeLongitude: number;
}

/**
 * What routing last said about a composition (§18, §20).
 *
 * Persisted so that opening a Trail tomorrow shows the same distance and the same line
 * as today. A road closure should not silently rewrite a journey someone saved.
 *
 * Provider-independent by construction: metres, seconds, GeoJSON and a bare provider
 * name. A raw Mapbox payload would make every saved Trail unreadable the day the
 * supplier changes.
 */
export interface TrailRouteSnapshot {
  readonly geometry: RouteGeometry;
  readonly bounds: RouteBounds;
  readonly distanceMeters: number;
  readonly durationSeconds: number;
  readonly provider: string;
  readonly calculatedAt: Date;
  /**
   * The trail revision this snapshot describes (§24).
   *
   * Equal to the Trail's revision means the line matches the current composition. A
   * boolean `dirty` would carry the same bit and none of the evidence.
   */
  readonly revision: number;
}

/** A → B with no stops, so the app can state what the whole detour costs (§39). */
export interface TrailBaseRoute {
  readonly distanceMeters: number;
  readonly durationSeconds: number;
  readonly calculatedAt: Date;
}

export interface Trail {
  readonly id: string;
  readonly ownerUserId: string;
  readonly status: TrailStatus;
  readonly origin: TrailEndpoint;
  readonly destination: TrailEndpoint;
  readonly stops: readonly TrailStop[];
  readonly revision: number;
  readonly route: TrailRouteSnapshot | null;
  readonly baseRoute: TrailBaseRoute | null;
  readonly createdAt: Date;
  readonly updatedAt: Date;
  readonly finalizedAt: Date | null;
}

/** Enough for a list row. Deliberately without geometry (§42, §80). */
export interface TrailSummary {
  readonly id: string;
  readonly status: TrailStatus;
  readonly originLabel: string | null;
  readonly destinationLabel: string | null;
  readonly stopCount: number;
  readonly distanceMeters: number | null;
  readonly durationSeconds: number | null;
  readonly revision: number;
  readonly updatedAt: Date;
}

/**
 * The most stops a Trail may hold (§14, §17).
 *
 * The binding constraint is the routing provider: the Mapbox Directions driving
 * profile accepts 25 coordinates per request, and origin and destination take two of
 * them, leaving 23. Fifteen sits deliberately below that ceiling — a margin for a
 * provider that publishes a smaller limit, and a UX bound besides, since a list of
 * twenty-three stops is no longer something a person reorders by dragging.
 *
 * Enforced in the application layer, which is where a caller gets a useful error.
 */
export const MAX_TRAIL_STOPS = 15;

/** True when the drawn route describes the composition as it now stands (§22, §24). */
export function isRouteCurrent(trail: Pick<Trail, 'revision' | 'route'>): boolean {
  return trail.route !== null && trail.route.revision === trail.revision;
}

/**
 * Extra distance and time the stops add, against the same endpoints with none (§38).
 *
 * Null when either snapshot is missing, rather than zero: "no detour" and "not
 * measured" are different claims, and rendering the second as the first would state a
 * fact Trilha does not have.
 */
export function trailDetour(
  trail: Pick<Trail, 'route' | 'baseRoute'>,
): { extraDistanceMeters: number; extraDurationSeconds: number } | null {
  const { route, baseRoute } = trail;
  if (route === null || baseRoute === null) return null;

  return {
    /* Clamped at zero: a provider can return a composed route marginally shorter than
       its own direct one, and a negative detour is not something to show a user. */
    extraDistanceMeters: Math.max(0, route.distanceMeters - baseRoute.distanceMeters),
    extraDurationSeconds: Math.max(0, route.durationSeconds - baseRoute.durationSeconds),
  };
}

File: apps/mobile/lib/features/trails/domain/trail.dart
import 'package:equatable/equatable.dart';

import '../../places/domain/place.dart';

/// A trail: what the user composed, not what the road network computed (§1).
///
/// Routing answers *how to get there*; a Trail records *what was chosen* — which
/// places, in which order, deliberately saved. The two are kept apart here for the
/// same reason they are on the server: a Trail outlives any particular calculation of
/// its route, and a route belongs to nobody.
class Trail extends Equatable {
  const Trail({
    required this.id,
    required this.status,
    required this.origin,
    required this.destination,
    required this.stops,
    required this.revision,
    required this.route,
    required this.baseRoute,
    required this.detour,
    required this.routeIsCurrent,
    required this.maxStops,
    required this.updatedAt,
  });

  final String id;
  final TrailStatus status;
  final TrailEndpoint origin;
  final TrailEndpoint destination;
  final List<TrailStop> stops;

  /// Send back as `expectedRevision` on the next mutation (§25, §57).
  final int revision;

  /// What routing last said. Null when the provider was unreachable at creation.
  final TrailRoute? route;
  final TrailMetrics? baseRoute;

  /// What the stops add over the same endpoints with none. Null when unmeasured.
  final TrailMetrics? detour;

  /// Whether [route] describes the composition as it now stands.
  final bool routeIsCurrent;

  /// The server's ceiling, so the UI never has to guess it (§14).
  final int maxStops;

  final DateTime updatedAt;

  bool get isFull => stops.length >= maxStops;

  bool containsPlace(String placeId) =>
      stops.any((TrailStop s) => s.placeId == placeId);

  /// A local reordering, for the drag preview (§56).
  ///
  /// Returns a Trail that looks like the drop already happened, without touching the
  /// server. The revision is unchanged on purpose: this is not a new version, it is a
  /// picture of one being proposed.
  ///
  /// Indices are post-removal, matching `ReorderableListView.onReorderItem` — the
  /// caller does not compensate for the dragged row leaving the list.
  Trail withStopsReordered(int oldIndex, int newIndex) {
    final List<TrailStop> next = <TrailStop>[...stops];
    final TrailStop moved = next.removeAt(oldIndex);
    next.insert(newIndex, moved);

    return Trail(
      id: id,
      status: status,
      origin: origin,
      destination: destination,
      stops: <TrailStop>[
        for (final (int index, TrailStop stop) in next.indexed)
          stop.atPosition(index + 1),
      ],
      revision: revision,
      route: route,
      baseRoute: baseRoute,
      detour: detour,
      routeIsCurrent: routeIsCurrent,
      maxStops: maxStops,
      updatedAt: updatedAt,
    );
  }

  @override
  List<Object?> get props => <Object?>[
    id,
    status,
    origin,
    destination,
    stops,
    revision,
    route,
    baseRoute,
    detour,
    routeIsCurrent,
    maxStops,
    updatedAt,
  ];
}

/// Lifecycle (§7).
///
/// `finalized` means "I finished composing this", never "I published it" — nobody
/// else can see a Trail. Editing a finalized Trail returns it to draft (§46).
enum TrailStatus {
  draft,
  finalized,
  archived;

  static TrailStatus fromCode(String code) => switch (code) {
    'FINALIZED' => TrailStatus.finalized,
    'ARCHIVED' => TrailStatus.archived,
    _ => TrailStatus.draft,
  };
}

class TrailEndpoint extends Equatable {
  const TrailEndpoint({required this.position, this.placeId, this.label});

  final LatLng position;
  final String? placeId;
  final String? label;

  @override
  List<Object?> get props => <Object?>[position, placeId, label];
}

/// A stop, always a resolved Place (§10).
class TrailStop extends Equatable {
  const TrailStop({
    required this.id,
    required this.placeId,
    required this.position,
    required this.source,
    required this.placeName,
    required this.placeCategoryId,
    required this.location,
  });

  final String id;
  final String placeId;

  /// 1-based visiting order.
  final int position;
  final TrailStopSource source;
  final String placeName;
  final String placeCategoryId;
  final LatLng location;

  TrailStop atPosition(int newPosition) => TrailStop(
    id: id,
    placeId: placeId,
    position: newPosition,
    source: source,
    placeName: placeName,
    placeCategoryId: placeCategoryId,
    location: location,
  );

  @override
  List<Object?> get props => <Object?>[
    id,
    placeId,
    position,
    source,
    placeName,
    placeCategoryId,
    location,
  ];
}

/// Where a stop came from. Provenance, never a quality signal (§11).
enum TrailStopSource {
  discovery,
  search,
  manual;

  String get code => switch (this) {
    TrailStopSource.discovery => 'DISCOVERY',
    TrailStopSource.search => 'SEARCH',
    TrailStopSource.manual => 'MANUAL',
  };

  static TrailStopSource fromCode(String code) => switch (code) {
    'DISCOVERY' => TrailStopSource.discovery,
    'SEARCH' => TrailStopSource.search,
    _ => TrailStopSource.manual,
  };
}

/// The stored route: the line and what it measured (§18).
class TrailRoute extends Equatable {
  const TrailRoute({
    required this.geometry,
    required this.bounds,
    required this.distanceMeters,
    required this.durationSeconds,
  });

  final List<LatLng> geometry;
  final MapBounds bounds;
  final int distanceMeters;
  final int durationSeconds;

  @override
  List<Object?> get props => <Object?>[
    geometry,
    bounds,
    distanceMeters,
    durationSeconds,
  ];
}

class TrailMetrics extends Equatable {
  const TrailMetrics({
    required this.distanceMeters,
    required this.durationSeconds,
  });

  final int distanceMeters;
  final int durationSeconds;

  @override
  List<Object?> get props => <Object?>[distanceMeters, durationSeconds];
}

/// A row in "Minhas trilhas" (§42, §68). No geometry, by design.
class TrailSummary extends Equatable {
  const TrailSummary({
    required this.id,
    required this.status,
    required this.stopCount,
    required this.revision,
    required this.updatedAt,
    this.originLabel,
    this.destinationLabel,
    this.distanceMeters,
    this.durationSeconds,
  });

  final String id;
  final TrailStatus status;
  final String? originLabel;
  final String? destinationLabel;
  final int stopCount;
  final int? distanceMeters;
  final int? durationSeconds;
  final int revision;
  final DateTime updatedAt;

  @override
  List<Object?> get props => <Object?>[
    id,
    status,
    originLabel,
    destinationLabel,
    stopCount,
    distanceMeters,
    durationSeconds,
    revision,
    updatedAt,
  ];
}

/// How a trail reads in a list: "Recife → João Pessoa".
String trailTitle(TrailSummary trail) {
  final String from = trail.originLabel ?? 'Origem';
  final String to = trail.destinationLabel ?? 'Destino';
  return '$from → $to';
}

/// "3 paradas · 145 km · 2h11", omitting whatever is not known yet.
String trailSubtitle(TrailSummary trail) {
  final List<String> parts = <String>[
    trail.stopCount == 1 ? '1 parada' : '${trail.stopCount} paradas',
  ];

  final int? distance = trail.distanceMeters;
  if (distance != null) parts.add(formatRouteDistance(distance));

  final int? duration = trail.durationSeconds;
  if (duration != null) parts.add(formatDuration(duration));

  return parts.join(' · ');
}

/// Formats a duration the way a driver would say it.
String formatDuration(int seconds) {
  if (seconds < 60) return '${seconds}s';

  final int totalMinutes = (seconds / 60).round();
  final int hours = totalMinutes ~/ 60;
  final int minutes = totalMinutes % 60;

  if (hours == 0) return '${minutes}min';
  if (minutes == 0) return '${hours}h';
  return '${hours}h${minutes.toString().padLeft(2, '0')}';
}

/// Formats a route distance, dropping precision as it grows.
String formatRouteDistance(int metres) {
  if (metres < 1000) return '$metres m';

  final double km = metres / 1000;
  return km < 100 ? '${km.toStringAsFixed(1)} km' : '${km.round()} km';
}
